### ISSUE TITLE
Threads stuck in Monitor.Wait in ContentFactory.InitField()

### ISSUE THREAD

**Issue by monsieurberry** [PiranhaCMS/piranha.core#1042]
Hi Hakan,

I have another (hopefully) quick question for you.

Occasionally in our production environment (hosted as an Azure App Service), the Thread Count jumps up quite dramatically on one of the instances and it becomes completely unresponsive until I go in and restart it manually. 

The application itself gets about 4-10k vistors a day.

I am running 8.0.0 and I have MemoryCache enabled and switched up to Full.

I was able to take a memory dump during one of these episodes and I found that quite a lot of these threads were hung up in Piranha.Services.ContentFactory.InitField(). Here is the Call stack of these threads starting from our controller up to where the thread seems to get locked.

```
System.Threading.ManualResetEventSlim.Wait(Int32, System.Threading.CancellationToken)+22d
--
System.Threading.Tasks.Task.SpinThenBlockingWait(Int32, System.Threading.CancellationToken)+92
System.Threading.Tasks.Task.InternalWaitCore(Int32, System.Threading.CancellationToken)+f1
System.Threading.Tasks.Task.Wait(Int32, System.Threading.CancellationToken)+35
Piranha.Services.ContentFactory.InitField(Microsoft.Extensions.DependencyInjection.IServiceScope, System.Object)+14e
Piranha.Services.ContentFactory.InitRegion(Microsoft.Extensions.DependencyInjection.IServiceScope, System.Object, Piranha.Models.RegionType)+bb
Piranha.Services.ContentFactory.Init[[System.__Canon, System.Private.CoreLib]](System.__Canon, Piranha.Models.ContentType)+c8
Piranha.Services.PageService+d__18`1[[System.__Canon, System.Private.CoreLib]].MoveNext()+13f
System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[[Piranha.Services.PageService+d__18`1[[System.__Canon, System.Private.CoreLib]], Piranha]](d__18`1 ByRef)+5e
System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1[[System.__Canon, System.Private.CoreLib]].Start[[Piranha.Services.PageService+d__18`1[[System.__Canon, System.Private.CoreLib]], Piranha]](d__18`1 ByRef)+2d
Piranha.Services.PageService.GetByIdAsync[[System.__Canon, System.Private.CoreLib]](System.Guid)+66
DigiBook.Controllers.HomeController+d__6.MoveNext()+74
System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[[DigiBook.Controllers.HomeController+d__6, DigiBook]](d__6 ByRef)+43
--
DigiBook.Controllers.HomeController.DefaultHomepageAsync(System.Guid, Boolean)+3c
```

I had a look in the InitField Function and can see the following line is performing a Wait() on the task. Is this something you think could potentially cause Deadlock? 

`Task.Run(async () => await ((Task)init.Invoke(field, param.ToArray())).ConfigureAwait(false)).Wait();`

I have to say, I am in no means an expert on these things and have been learning as much as I can over the last few weeks. I am just trying to get to the bottom of this issue as it's worrisome that it happens for no reason.

I can provide you with more information if required.

Many Thanks

Adam

**Comment by tidyui** [comment]
Hi there! As we haven't seen this issue before, do you have any **custom fields** in you application performing actions inside of their `Init` methods that could potentially block the application?

Regards

**Comment by monsieurberry** [comment]
Hi Hakan,

Thanks for getting back to me. Yes I have three custom fields but they are all pretty basic in terms of their Init() methods. I will paste them below for reference.

**Currency Field**
```
using Piranha.Extend;

namespace DigiBook.Cms.Models.Fields
{
    [FieldType(Name = "Currency Field", Component = "currency-field")]
    public class CurrencyField : IField
    {
        public string Value { get; set; }

        public string GetTitle()
        {
            return Value;
        }

        public void Init()
        {
            // Default to Euro
            if (Value == null)
                Value = "C10B9B08-A6FE-47ED-BAA8-9B7CC40422D0";
        }
    }
}
```

**Guid Field**
```
using Piranha.Extend;
using System;

namespace DigiBook.Cms.Models.Fields
{
    [FieldType(Name = "Guid Field", Component = "guid-field")]
    public class GuidField : IField
    {
        public string Value { get; set; }

        public string GetTitle()
        {
            return Value;
        }

        public void Init()
        {
            if (string.IsNullOrEmpty(Value))
                Value = Guid.NewGuid().ToString();
        }
    }
}
```

**Provider Type Field**
```
using Piranha.Extend;

namespace DigiBook.Cms.Models.Fields
{
    [FieldType(Name = "Provider Type Field", Component = "provider-type-field")]
    public class ProviderTypeField : IField
    {
        public string Value { get; set; }

        public string GetTitle()
        {
            return Value;
        }

        public void Init()
        {
            // Default to Hotel
            // TODO HOOK THIS INTO API
            if (Value == null)
                Value = "E9BBA8AD-8CF2-45D2-9155-37ABECED25C8";
        }
    }
}
```

**Comment by monsieurberry** [comment]
Hi again, Hakan,

I just read an interesting article about Blocking with async code

https://medium.com/rubrikkgroup/understanding-async-avoiding-deadlocks-e41f8f2c6f5d

From what I gather a call to Wait() blocks the thread. Could it be possible that under heavy load the threads are all getting blocked and piling up?

Thanks

Adam

**Comment by tidyui** [comment]
Hi there! The whole point when running an async method synchronous is that it should wait in the thread just like a synchronous operation would've done. There's more information here about the preferred solutions to doing this:

https://docs.microsoft.com/en-us/archive/msdn-magazine/2015/july/async-programming-brownfield-async-development

We have chosen the second solution where the method is actually dispatched on the thread pool. The best way from a resource point of view would be to use `await` on the call, the reason for the code looking the way it does is simply that the whole library started out being sync only and it had to be able to dispatch an asynchronous method.

**Comment by monsieurberry** [comment]
Thanks Hakan,

Do you think it will be changed to use await in the future?

Having done a bit more research this evening I was advised to install a nuget package called 'Microsoft.VisualStudio.Threading.Analyzers'. After installing this and looking at that line of code, the Wait() call is highlighted and I get a roslyn style code warning stating "Synchronsously waiting on tasks or awaiters may cause deadlocks. Use await or JoinableTaskFactory.Run instead".

This has me almost convinced that the reason things are spinning out of control is because of this piece of code, especially seeing as the threads are all hanging in that particular function. It's kind of freaking me out a small bit as it's like a ticking timebomb in our production code which can only be mitigated by restarting the App Service

Any help you could give would be greatly appreciated :-)

Thanks!

Adam

![image](https://user-images.githubusercontent.com/706856/74275781-6b197c00-4d0c-11ea-89f7-03ae1fdd49d8.png)

**Comment by tidyui** [comment]
I have updated the code to make `ContentFactory.InitField` async, which was a horrendous task 😂 (as you can see from the commit https://github.com/PiranhaCMS/piranha.core/commit/d0147eb685b2694223cf2e520c1388cc59ec883e).

As you probably need a **lot** of requests to encounter thread starvation I haven't been able to verify if the change solves the problem, but it should as the entire flow is now completely async. Would you be interested in testing beta packages of `8.1` to see if this resolves your issue?

Best regards

**Comment by monsieurberry** [comment]
Hi Hakan,

I can't tell you how grateful I am for you doing this for me. It does indeed look like a mammoth task so thanks so much for looking at it so quickly.

I would be delighted to beta test 8.1 for you. Just let me know how I can get hold of the 8.1 packages and I'll promote to our prod environment as soon as I can arrange it.

Kind Regards

Adam

**Comment by tidyui** [comment]
How does your build pipeline look, i.e do you need public packages on NuGet for any automated builds to work or can you build and deploy with local NuGet-packages?

**Comment by monsieurberry** [comment]
Hi Hakan,

We don;t currently have an automated build pipeline. I'm doing everything manually in Visual Studio and publishing from there (new system so not explored automation yet).

In theory I probably could use local NuGet packages. I've never done it before but I'm sure I could get my head around it. Is it possible to add the package as pre-release like in the attached image - or would you rather not go down this route? 

Whatever works best for you. I should be able to just install the package and publish.

Thanks

Adam

![image](https://user-images.githubusercontent.com/706856/74325305-44952880-4d80-11ea-9e23-bcc5fd85a4a3.png)

**Comment by tidyui** [comment]
I have merged the code into master and updated project & nuspec versions to `8.1.0-beta4` which is our internal pre-release number. You can build the packages locally by:

1. Pull the latest source code
2. Run `npm-install` in `core/Piranha.Manager` from the command line
3. Run `./pack.sh` in the root directory from the command line

This will give you all of the packages in an `artifacts` folder in the project root.

Please note that there will be some differences you'll have to make, for example you'll have to include a database-provider package due to changes in EF Core Migrations >= 3.0. Take a look at the examples in the repository for guidance.

Best regards

**Comment by monsieurberry** [comment]
That's great thanks Hakan.

I'm away on business tomorrow but I'll have a look on Friday with a view to pushing live on Monday next week.

Kind Regards

Adam

**Comment by monsieurberry** [comment]
Quick update for you, Hakan.

I was able to build and install the 8.1 packages and push to our test environment successfully. Thank you for your instructions.

It’s been through QA testing today and nothing out of the ordinary turned up, so we will look to push to prod on Monday. From then on I’ll keep you posted on how things are.

The issue was very sporadic but usually happened in high traffic. We should have a good indication of whether it’s fixed or not within the week but as you stated in a previous comment, I can see no reason why the fix you have made wouldn’t resolve the issue.

Thanks again
Adam

**Comment by monsieurberry** [comment]
Hi Hakan,

Just to let you know this is all in production now and we haven't seen any occurences of the issue yet. I'll update you after a few days of heavy traffic.

One thing I have noticed with 8.1 (I can log a separate ticket for this if required) Is that our blog seems to be broken.

When I view the archive or try and load the archive page in the manage I get the following error.

`InvalidOperationException: A second operation started on this context before a previous operation completed. This is usually caused by different threads using the same instance of DbContext. For more information on how to avoid threading issues with DbContext`

Here's where I define EF in Configure Services.

```
services.AddPiranhaEF<SQLServerDb>(options =>
  options.UseSqlServer([ConnectionString])
);
```

Any ideas what I could be doing wrong?
Thanks
Adam

**Comment by tidyui** [comment]
Awesome! For your other issue, activate MARS on the connection string to resolve it!

**Comment by monsieurberry** [comment]
Didn't know about MARS. I added it to the connectionstring but it made no difference...?

When I switch back to my other branch on 8.0 it seems to work OK. I know I'm missing something somewhere I just can't figure out what.

It works fine in the sample project but I know that's working with SqlLite so maybe I'm missing something else on the config?

Thanks!

**Comment by monsieurberry** [comment]
Hi Hakan,

Seems to work OK if I change the EF registering to transient...

```
services.AddPiranhaEF<SQLServerDb>(options =>
    options.UseSqlServer(Configuration.GetConnectionString("DigiBookCMS")),
    ServiceLifetime.Transient
);
```

If I do this will it mess up anything else in Piranha do you think?

Ta

Adam

**Comment by tidyui** [comment]
Since transient means that a new `DbContext` should be created every time it's requested instead once per request (Scoped) it will surely add overhead since more objects need to be instantiated and more connections need to opened and closed. 

Just to clarify. This **only** happens when opening the archive view in the manager?

**Comment by monsieurberry** [comment]
Yeah I thought as much. Probably not a wise idea.

Without that transient change it happens when viewing the archive through the front end and also through the manager. With the change it only happens in the manager.

Thanks

Adam

**Comment by tidyui** [comment]
And could you post a full stack trace, maybe there's an indication on which line the error occurs.

**Comment by tidyui** [comment]
And when viewing a single post?

**Comment by monsieurberry** [comment]
Yes of course.

**Viewing Archive Page**
`at Microsoft.EntityFrameworkCore.Internal.ConcurrencyDetector.EnterCriticalSection() at Microsoft.EntityFrameworkCore.Query.Internal.QueryingEnumerable`1.AsyncEnumerator.MoveNextAsync() at Microsoft.EntityFrameworkCore.Query.ShapedQueryCompilingExpressionVisitor.SingleOrDefaultAsync[TSource](IAsyncEnumerable`1 asyncEnumerable, CancellationToken cancellationToken) at Microsoft.EntityFrameworkCore.Query.ShapedQueryCompilingExpressionVisitor.SingleOrDefaultAsync[TSource](IAsyncEnumerable`1 asyncEnumerable, CancellationToken cancellationToken) at Piranha.Repositories.PageRepository.GetById[T](Guid id) at Piranha.Services.PageService.GetByIdAsync[T](Guid id) at Piranha.Services.PostService.GetByIdAsync[T](Guid id, IList`1 blogPages) at Piranha.Services.ArchiveService.GetByIdAsync[T](Guid archiveId, Nullable`1 currentPage, Nullable`1 categoryId, Nullable`1 tagId, Nullable`1 year, Nullable`1 month, Nullable`1 pageSize) at DigiBook.Controllers.BlogController.BlogAsync(Guid id, Nullable`1 year, Nullable`1 month, Nullable`1 page, Nullable`1 category, Nullable`1 tag) in C:\Projects\DigiBook\DigiBook\Controllers\BlogController.cs:line 54 at Microsoft.AspNetCore.Mvc.Infrastructure.ActionMethodExecutor.TaskOfIActionResultExecutor.Execute(IActionResultTypeMapper mapper, ObjectMethodExecutor executor, Object controller, Object[] arguments) at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.<InvokeActionMethodAsync>g__Logged|12_1(ControllerActionInvoker invoker) at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.<InvokeNextActionFilterAsync>g__Awaited|10_0(ControllerActionInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted) at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.Rethrow(ActionExecutedContextSealed context) at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted) at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.<InvokeInnerFilterAsync>g__Awaited|13_0(ControllerActionInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted) at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeNextExceptionFilterAsync>g__Awaited|25_0(ResourceInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)`

**Opening Archive page in Manager - Blank screen with 500 error in console window. Clicking url gives...**
```
Microsoft.EntityFrameworkCore.Internal.ConcurrencyDetector.EnterCriticalSection()
Microsoft.EntityFrameworkCore.Query.Internal.QueryingEnumerable<T>+AsyncEnumerator.MoveNextAsync()
Microsoft.EntityFrameworkCore.Query.ShapedQueryCompilingExpressionVisitor.SingleOrDefaultAsync<TSource>(IAsyncEnumerable<TSource> asyncEnumerable, CancellationToken cancellationToken)
Microsoft.EntityFrameworkCore.Query.ShapedQueryCompilingExpressionVisitor.SingleOrDefaultAsync<TSource>(IAsyncEnumerable<TSource> asyncEnumerable, CancellationToken cancellationToken)
Piranha.Repositories.PageRepository.GetById<T>(Guid id)
Piranha.Services.PageService.GetByIdAsync<T>(Guid id)
Piranha.Services.PostService.GetByIdAsync<T>(Guid id, IList<PageInfo> blogPages)
Piranha.Services.PostService.GetAllAsync<T>(Guid blogId, Nullable<int> index, Nullable<int> pageSize)
Piranha.Manager.Services.PostService.GetList(Guid archiveId, int index)
Piranha.Manager.Controllers.PostApiController.List(Guid id, int index)
lambda_method(Closure , object )
Microsoft.Extensions.Internal.ObjectMethodExecutorAwaitable+Awaiter.GetResult()
Microsoft.AspNetCore.Mvc.Infrastructure.ActionMethodExecutor+AwaitableObjectResultExecutor.Execute(IActionResultTypeMapper mapper, ObjectMethodExecutor executor, object controller, object[] arguments)
System.Threading.Tasks.ValueTask<TResult>.get_Result()
System.Runtime.CompilerServices.ValueTaskAwaiter<TResult>.GetResult()
Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.<InvokeActionMethodAsync>g__Logged|12_1(ControllerActionInvoker invoker)
Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.<InvokeNextActionFilterAsync>g__Awaited|10_0(ControllerActionInvoker invoker, Task lastTask, State next, Scope scope, object state, bool isCompleted)
Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.Rethrow(ActionExecutedContextSealed context)
Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.Next(ref State next, ref Scope scope, ref object state, ref bool isCompleted)
Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.<InvokeInnerFilterAsync>g__Awaited|13_0(ControllerActionInvoker invoker, Task lastTask, State next, Scope scope, object state, bool isCompleted)
Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeNextResourceFilter>g__Awaited|24_0(ResourceInvoker invoker, Task lastTask, State next, Scope scope, object state, bool isCompleted)
Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.Rethrow(ResourceExecutedContextSealed context)
Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.Next(ref State next, ref Scope scope, ref object state, ref bool isCompleted)
Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeFilterPipelineAsync>g__Awaited|19_0(ResourceInvoker invoker, Task lastTask, State next, Scope scope, object state, bool isCompleted)
Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeAsync>g__Logged|17_1(ResourceInvoker invoker)
Microsoft.AspNetCore.Routing.EndpointMiddleware.<Invoke>g__AwaitRequestTask|6_0(Endpoint endpoint, Task requestTask, ILogger logger)
Microsoft.AspNetCore.Authorization.AuthorizationMiddleware.Invoke(HttpContext context)
Microsoft.AspNetCore.Authentication.AuthenticationMiddleware.Invoke(HttpContext context)
Piranha.AspNetCore.SitemapMiddleware.Invoke(HttpContext context, IApi api, IApplicationService service)
Piranha.AspNetCore.StartPageMiddleware.Invoke(HttpContext context, IApi api, IApplicationService service)
Piranha.AspNetCore.ArchiveMiddleware.Invoke(HttpContext context, IApi api, IApplicationService service)
Piranha.AspNetCore.PostMiddleware.Invoke(HttpContext context, IApi api, IApplicationService service)
Piranha.AspNetCore.PageMiddleware.Invoke(HttpContext context, IApi api, IApplicationService service)
Piranha.AspNetCore.AliasMiddleware.Invoke(HttpContext context, IApi api, IApplicationService service)
Piranha.AspNetCore.ApplicationMiddleware.Invoke(HttpContext context, IApi api, IApplicationService service)
Microsoft.AspNetCore.Localization.RequestLocalizationMiddleware.Invoke(HttpContext context)
Microsoft.AspNetCore.Session.SessionMiddleware.Invoke(HttpContext context)
Microsoft.AspNetCore.Session.SessionMiddleware.Invoke(HttpContext context)
Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddleware.Invoke(HttpContext context)
```

**Viewing single blog post from front-end**

`at Microsoft.EntityFrameworkCore.Internal.ConcurrencyDetector.EnterCriticalSection() at Microsoft.EntityFrameworkCore.Query.Internal.QueryingEnumerable`1.AsyncEnumerator.MoveNextAsync() at Microsoft.EntityFrameworkCore.Query.ShapedQueryCompilingExpressionVisitor.SingleOrDefaultAsync[TSource](IAsyncEnumerable`1 asyncEnumerable, CancellationToken cancellationToken) at Microsoft.EntityFrameworkCore.Query.ShapedQueryCompilingExpressionVisitor.SingleOrDefaultAsync[TSource](IAsyncEnumerable`1 asyncEnumerable, CancellationToken cancellationToken) at Piranha.Repositories.PageRepository.GetById[T](Guid id) at Piranha.Services.PageService.GetByIdAsync[T](Guid id) at Piranha.Services.PostService.GetByIdAsync[T](Guid id, IList`1 blogPages) at DigiBook.Controllers.BlogController.BlogPostAsync(Guid id) in C:\Projects\DigiBook\DigiBook\Controllers\BlogController.cs:line 78 at Microsoft.AspNetCore.Mvc.Infrastructure.ActionMethodExecutor.TaskOfIActionResultExecutor.Execute(IActionResultTypeMapper mapper, ObjectMethodExecutor executor, Object controller, Object[] arguments) at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.<InvokeActionMethodAsync>g__Logged|12_1(ControllerActionInvoker invoker) at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.<InvokeNextActionFilterAsync>g__Awaited|10_0(ControllerActionInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted) at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.Rethrow(ActionExecutedContextSealed context) at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted) at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.<InvokeInnerFilterAsync>g__Awaited|13_0(ControllerActionInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted) at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeNextExceptionFilterAsync>g__Awaited|25_0(ResourceInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)`

Hope these are useful and thank you
Adam

**Comment by tidyui** [comment]
Hi again. I found the error debugging locally. In the whole refactoring commit of awaiting `InitField` there was **one** async call that was not being awaited which caused the errors for Posts. We will have to push an update for `Piranha.Data.EF`.

**Comment by monsieurberry** [comment]
That's great thanks Hakan, you're a star.

In retrospect I probably should have opened another issue for this. I thought it was something silly I had done in the config.

I can do any tests for you if you need me to, just let me know :-)

**Comment by tidyui** [comment]
Nah it was our fault. We usually don't add such big chances as this issue this close to release, so the issue slipped us by.

**Comment by tidyui** [comment]
`Piranha.Data.EF.SQLServer 8.1.1` should be available on NuGet now.

**Comment by monsieurberry** [comment]
Works beautifully!

Thanks Hakan

**Comment by monsieurberry** [comment]
Thanks Hakan,

Meant to get back to you and say that this definitely seems to have solved the problem.

Thanks again for looking at it so quickly. Much appreciated

Adam
